vulnerabilities/scraper/ruby.py

```diff
@@ -0,0 +1,104 @@
+            DOWNLOAD_PATH, 'ruby.zip'))
+    ZipFile(path_to_zip).extractall(DOWNLOAD_PATH)
+    os.remove(path_to_zip)
+
+
+def path_of_yaml_of_all_packages():
+    gem_path = os.path.join(DOWNLOAD_PATH, 'ruby-advisory-db-master', 'gems')
+    rubies_path = os.path.join(
+        DOWNLOAD_PATH,
+        'ruby-advisory-db-master',
+        'rubies')
+    for (
+            package_path,
+            _,
+            yaml_names) in chain(
+            os.walk(gem_path),
+            os.walk(rubies_path)):
+        for yaml_name in yaml_names:
+            yield os.path.join(package_path, yaml_name)
+
+
+def get_all_versions_of_package(package_name):
+    url_to_load = 'https://rubygems.org/api/v1/versions/' + package_name + '.yaml'
+    try:
+        page = urllib.request.urlopen(url_to_load)
+        package_history = saneyaml.load(page)
+    except HTTPError:
+        return []
+    for version in package_history:
+        yield version['number']
+
+
+def get_patched_range(spec_list):
+    if spec_list:
+        def remove_space(string): return string.replace(' ', '')
+        spec_list = list(map(remove_space, spec_list))
+        for spec in spec_list:
+            if 'rc' in spec:
+                continue
+            yield RangeSpecifier(spec)
+    else:
+        return None
+
+
+def import_vulnerabilities():
+    get_rubycve_db()
+    ids = set()
+    vulnerability_to_package_map = []
+    for vulnerability_path in path_of_yaml_of_all_packages():
+        with open(vulnerability_path) as yaml_file:
+            vulnerability = saneyaml.load(yaml_file)
+            package_name = vulnerability.get(
+                'engine', vulnerability.get('gem'))
+            summary = vulnerability.get('description', '')
+            if 'cve' in vulnerability:
+                vulnerability_id = 'CVE-{}'.format(vulnerability['cve'])
+                if vulnerability_id in ids:
+                    continue
+                ids.add(vulnerability_id)
+            else:
+                continue
+            severity = vulnerability.get(
+                'cvss_v3', vulnerability.get('cvss_v2'))
+            advisory_url = vulnerability.get('url')
+            specs = list(
+                get_patched_range(
+                    vulnerability.get('patched_versions')))
+            all_versions = set(list(get_all_versions_of_package(package_name)))
+            unaffected_versions = set()
+            if specs:
+                for version in all_versions:
+                    for spec in specs:
+                        if version in spec:
+                            unaffected_versions.add(version)
+                            break
+            affected_versions = all_versions - unaffected_versions
+            vulnerability_to_package_map.append({
+                'package_name': package_name,
+                'summary': summary,
+                'cve_id': vulnerability_id,
+                'fixed_versions': unaffected_versions,
```

Comment:
I think this came up elsewhere before: Fixed (or patched) and unaffected packages are not the same thing. The rubysec yaml schema includes both:
```
$ grep -A 2 _versions gems/nokogiri/CVE-2015-5312.yml
unaffected_versions:
  - "< 1.6.0"

patched_versions:
  - ">= 1.6.7.1"
```
With the current data model of vulnerablecode, unaffected versions would just be packages that have neither a corresponding `ImpactedPackage` nor a `ResolvedPackage`. I don't know off the top of my head whether we make this distinction elsewhere, but this is one reason why we need a better architecture for importers: To have one place where those decisions are made.

Reply:
Exactly, that would need some refactoring for all the current importers.